File: Medium/1382. Balance a Binary Search Tree.java
class Solution1 {
    public TreeNode balanceBST(TreeNode root) {
        final var nodes = new ArrayList<TreeNode>();
        dfs(root, nodes);
        final int n = nodes.size();
        var newRoot = nodes.get(n / 2);
        newRoot.left = null;
        newRoot.right = null;
        var dummy1 = newRoot;
        for (int i = n / 2 - 1; i >= 0; i--) {
            dummy1.left = nodes.get(i);
            dummy1 = dummy1.left;
            dummy1.right = null;
        }
        dummy1.left = null;
        var dummy2 = newRoot;
        for (int i = n / 2 + 1; i < n; i++) {
            dummy2.right = nodes.get(i);
            dummy2 = dummy2.right;
            dummy2.left = null;
        }
        dummy2.right = null;
        return newRoot;
    }

    private void dfs(TreeNode node, List<TreeNode> nodes) {
        if (node == null) {
            return;
        }
        dfs(node.left, nodes);
        nodes.add(node);
        dfs(node.right, nodes);
    }
}

class Solution {
    public TreeNode balanceBST(TreeNode root) {
        final var nodes = new ArrayList<TreeNode>();
        dfs(root, nodes);
        return buildBalancedBST(nodes, 0, nodes.size() - 1);
    }

    private TreeNode buildBalancedBST(List<TreeNode> nodes, int start, int end) {
        if (start > end) {
            return null;
        }
        int mid = (start + end) / 2;
        TreeNode node = nodes.get(mid);
        node.left = buildBalancedBST(nodes, start, mid - 1);
        node.right = buildBalancedBST(nodes, mid + 1, end);
        return node;
    }

    private void dfs(TreeNode node, List<TreeNode> nodes) {
        if (node == null) {
            return;
        }
        dfs(node.left, nodes);
        nodes.add(node);
        dfs(node.right, nodes);
    }
}

File: Medium/1382. Balance a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List, Optional

from TreeNode import TreeNode


class Solution:
    def inorder(self, node: Optional[TreeNode], vals: List[int]) -> None:
        if not node:
            return
        self.inorder(node.left, vals)
        vals.append(node.val)
        self.inorder(node.right, vals)
    def build(self, vals: List[int], l: int, r: int) -> Optional[TreeNode]:
        if l > r:
            return None
        m = (l + r) // 2
        # node = TreeNode(vals[mid])
        # node.left = self.build(vals, l, mid - 1)
        # node.right = self.build(vals, mid + 1, r)
        # return node
        return TreeNode(vals[m], self.build(vals, l, m - 1), self.build(vals, m + 1, r))
    def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        vals = []
        self.inorder(root, vals)
        return self.build(vals, 0, len(vals) - 1)